clang/lib/Sema/SemaChecking.cpp

```diff
@@ -179,6 +179,56 @@ static bool SemaBuiltinCHERICapCreate(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+static bool checkBuiltinArgument(Sema &S, CallExpr *E, unsigned ArgIndex);
+
+static bool checkCapArg(Sema &S, CallExpr *TheCall, unsigned ArgIndex,
+                        QualType &SrcTy) {
+  clang::Expr *Source = TheCall->getArg(ArgIndex);
+  SrcTy = Source->getType();
+  // We should also be able to use it with arrays (TODO: and functions?).
+  if (SrcTy->canDecayToPointerType() && SrcTy->isArrayType()) {
+    SrcTy = S.Context.getDecayedType(SrcTy);
+  }
+  if (!SrcTy->isCapabilityPointerType() && !SrcTy->isIntCapType()) {
+    S.Diag(Source->getExprLoc(), diag::err_typecheck_expect_capability_operand)
+        << SrcTy;
+    return true;
+  }
+  ExprResult SrcArg = S.PerformCopyInitialization(
+      InitializedEntity::InitializeParameter(S.Context, SrcTy, false),
+      SourceLocation(), Source);
+  if (SrcArg.isInvalid())
+    return true;
+  TheCall->setArg(ArgIndex, SrcArg.get());
+  return false;
+}
+
+/// The CHERI accessors should accept both pointer types and (u)intcap_t.
+static bool SemaBuiltinCheriGet(Sema &S, CallExpr *TheCall) {
```

Comment:
Just inline this? It's only used once, and you have other inlined checks of similar complexity.

Reply:
Yeah I originally had a lot more code there but after refactoring the shared stuff it doesn't need to be a function.